**Pandas** — это библиотека Python, предоставляющая широкие возможности для анализа данных. 

Основными структурами данных в Pandas являются классы **Series** (*одномерный индексированный массив данных некоторого фиксированного типа*) и **DataFrame** (*двухмерная структура данных, представляющая собой таблицу, каждый столбец которой содержит данные одного типа; можно представлять её как словарь объектов типа Series*).

Структура DataFrame отлично подходит для представления реальных данных: строки соответствуют признаковым описаниям отдельных объектов, а столбцы соответствуют признакам.

In [77]:
import pandas as pd
import numpy as np

Анализируется набор данных по оттоку клиентов телеком-оператора. С помощью функции `read_csv('file_path')` считываются данные из файла формата .csv. Метод `head()` позволяет посмотреть первые 5 строк данных таблицы.

In [78]:
df = pd.read_csv('telecom_churn.csv')
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

Свойство (*property*) `shape` класса DataFrame возвращает кортеж, представляющий размерность данных.

In [79]:
df.shape

(3333, 20)

Свойство `columns` возвращает объект типа **Index**, в котором перечислены имена всех столбцов (названия признаков).

In [80]:
df.columns

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

Метод `info()` выводит общую информацию о объекте класса DataFrame – размерность, названия столбцов, их типы, ненулевые значения и объем занимаемой памяти.  

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
State                     3333 non-null object
Account length            3333 non-null int64
Area code                 3333 non-null int64
International plan        3333 non-null object
Voice mail plan           3333 non-null object
Number vmail messages     3333 non-null int64
Total day minutes         3333 non-null float64
Total day calls           3333 non-null int64
Total day charge          3333 non-null float64
Total eve minutes         3333 non-null float64
Total eve calls           3333 non-null int64
Total eve charge          3333 non-null float64
Total night minutes       3333 non-null float64
Total night calls         3333 non-null int64
Total night charge        3333 non-null float64
Total intl minutes        3333 non-null float64
Total intl calls          3333 non-null int64
Total intl charge         3333 non-null float64
Customer service calls    3333 non-null int64


Метод `astype('dtype')` позволяет изменить тип колонки к указанному типу.

In [82]:
df['Churn'] = df['Churn'].astype('int64')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
State                     3333 non-null object
Account length            3333 non-null int64
Area code                 3333 non-null int64
International plan        3333 non-null object
Voice mail plan           3333 non-null object
Number vmail messages     3333 non-null int64
Total day minutes         3333 non-null float64
Total day calls           3333 non-null int64
Total day charge          3333 non-null float64
Total eve minutes         3333 non-null float64
Total eve calls           3333 non-null int64
Total eve charge          3333 non-null float64
Total night minutes       3333 non-null float64
Total night calls         3333 non-null int64
Total night charge        3333 non-null float64
Total intl minutes        3333 non-null float64
Total intl calls          3333 non-null int64
Total intl charge         3333 non-null float64
Customer service calls    3333 non-null int64


Метод `describe()` выводит основные статистические характеристики данных по каждому числовому признаку (типы `int64` и `float64`): число non-NA (*None, NaN, NaT*) значений, среднее значение, СКО и квартили.

In [83]:
df.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls        Churn  
count        3333.000000             3333.000000  3333.000000  
mean            2.764581                1.562856     0.144914  
std             0.753773                1.315491     0.352067  
min             0.000000                0.000000     0.000000  
25%             2.300000                1.000000     0.000000  
50%             2.780000                1.000000     0.000000  
75%             3.270000                2.000000     0.000000  
max             5.400000                9.000000     1.000000

Чтобы посмотреть статистику по нечисловым признакам, нужно явно указать интересующие нас типы в параметре `include`.

In [84]:
df.describe(include=['object', 'bool'])

State International plan Voice mail plan
count   3333               3333            3333
unique    51                  2               2
top       WV                 No              No
freq     106               3010            2411

Метод `value_counts` возвращает количество уникальных значений (в данном случае смотрим на распределение данных по признаку `Churn`).

In [85]:
df['Churn'].value_counts()

0    2850
1     483
Name: Churn, dtype: int64

Указав значение параметра `normalize=True`, можно посмотреть относительные частоты (в данном примере смотрим на распределение пользователей по параметру `Area code`).

In [86]:
df['Area code'].value_counts(normalize=True)

415    0.496550
510    0.252025
408    0.251425
Name: Area code, dtype: float64

### Сортировка

DataFrame можно отсортировать по значению какого-нибудь из признаков (в данном примере сортировка осуществляется по признаку `Total day charge`; параметр `ascending=False` означает сортировку по убыванию).

In [87]:
df.sort_values(by='Total day charge', ascending=False)

State  Account length  Area code International plan Voice mail plan  \
365     CO             154        415                 No              No   
985     NY              64        415                Yes              No   
2594    OH             115        510                Yes              No   
156     OH              83        415                 No              No   
605     MO             112        415                 No              No   
...    ...             ...        ...                ...             ...   
1986    WI              70        415                 No              No   
2753    OH             134        415                 No              No   
2736    OK             127        510                 No             Yes   
1397    VT             101        510                 No              No   
1345    SD              98        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
365                       0              350.8               75   
985                       0              346.8               55   
2594                      0              345.3               81   
156                       0              337.4              120   
605                       0              335.5               77   
...                     ...                ...              ...   
1986                      0                7.9              100   
2753                      0                7.8               86   
2736                     27                2.6              113   
1397                      0                0.0                0   
1345                      0                0.0                0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
365              59.64              216.5               94             18.40   
985              58.96              249.5               79             21.21   
2594             58.70              203.4              106             17.29   
156              57.36              227.4              116             19.33   
605              57.04              212.5              109             18.06   
...                ...                ...              ...               ...   
1986              1.34              136.4               83             11.59   
2753              1.33              171.4              100             14.57   
2736              0.44              254.0              102             21.59   
1397              0.00              192.1              119             16.33   
1345              0.00              159.6              130             13.57   

      Total night minutes  Total night calls  Total night charge  \
365                 253.9                100               11.43   
985                 275.4                102               12.39   
2594                217.5                107                9.79   
156                 153.9                114                6.93   
605                 265.0                132               11.93   
...                   ...                ...                 ...   
1986                156.6                 89                7.05   
2753                186.5                 80                8.39   
2736                242.7                156               10.92   
1397                168.8                 95                7.60   
1345                167.1                 88                7.52   

      Total intl minutes  Total intl calls  Total intl charge  \
365                 10.1                 9               2.73   
985                 13.3                 9               3.59   
2594                11.8                 8               3.19   
156                 15.8                 7               4.27   
605                 12.7                 8               3.43   
...                  ...               ...                ...   
1986                12.1                 1            

Сортировать можно и по группе столбцов, причем для каждого столбца можно указать свой порядок сортировки. 

In [88]:
df.sort_values(by=['Churn', 'Total day charge'], ascending=[True, False]).head()

State  Account length  Area code International plan Voice mail plan  \
688     MN              13        510                 No             Yes   
2259    NC             210        415                 No             Yes   
534     LA              67        510                 No              No   
575     SD             114        415                 No             Yes   
2858    AL             141        510                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
688                      21              315.6              105   
2259                     31              313.8               87   
534                       0              310.4               97   
575                      36              309.9               90   
2858                     28              308.0              123   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
688              53.65              208.9               71             17.76   
2259             53.35              147.7              103             12.55   
534              52.77               66.5              123              5.65   
575              52.68              200.3               89             17.03   
2858             52.36              247.8              128             21.06   

      Total night minutes  Total night calls  Total night charge  \
688                 260.1                123               11.70   
2259                192.7                 97                8.67   
534                 246.5                 99               11.09   
575                 183.5                105                8.26   
2858                152.9                103                6.88   

      Total intl minutes  Total intl calls  Total intl charge  \
688                 12.1                 3               3.27   
2259                10.1                 7               2.73   
534                  9.2                10               2.48   
575                 14.2                 2               3.83   
2858                 7.4                 3               2.00   

      Customer service calls  Churn  
688                        3      0  
2259                       3      0  
534                        4      0  
575                        1      0  
2858                       1      0

### Индексация и извлечение данных

Для извлечения отдельного столбца можно использовать кострукцию вида `DataFrame['Name']`. 

Вопрос: **Какова доля нелояльных пользователей в нашем датафрейме?**

👇

In [89]:
df['Churn'].mean()

0.14491449144914492

**Логическая индексация** DataFrame по одному столбцу выглядит следующим образом: `df[P(df['Name'])]`, где `P(df['Name'])` – это некоторое логическое условие, проверяемое для каждого элемента столбца `Name`. Итогом такой индексации является DataFrame, состоящий только из строк, удовлетворяющих условию `P` по столбцу `Name`.

In [90]:
df[df['Churn'] == 1]

State  Account length  Area code International plan Voice mail plan  \
10      IN              65        415                 No              No   
15      NY             161        415                 No              No   
21      CO              77        408                 No              No   
33      AZ              12        408                 No              No   
41      MD             135        408                Yes             Yes   
...    ...             ...        ...                ...             ...   
3301    CA              84        415                 No              No   
3304    IL              71        510                Yes              No   
3320    GA             122        510                Yes              No   
3322    MD              62        408                 No              No   
3323    IN             117        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
10                        0              129.1              137   
15                        0              332.9               67   
21                        0               62.4               89   
33                        0              249.6              118   
41                       41              173.1               85   
...                     ...                ...              ...   
3301                      0              280.0              113   
3304                      0              186.1              114   
3320                      0              140.0              101   
3322                      0              321.1              105   
3323                      0              118.4              126   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
10               21.95              228.5               83             19.42   
15               56.59              317.8               97             27.01   
21               10.61              169.9              121             14.44   
33               42.43              252.4              119             21.45   
41               29.43              203.9              107             17.33   
...                ...                ...              ...               ...   
3301             47.60              202.2               90             17.19   
3304             31.64              198.6              140             16.88   
3320             23.80              196.4               77             16.69   
3322             54.59              265.5              122             22.57   
3323             20.13              249.3               97             21.19   

      Total night minutes  Total night calls  Total night charge  \
10                  208.8                111                9.40   
15                  160.6                128                7.23   
21                  209.6                 64                9.43   
33                  280.2                 90               12.61   
41                  122.2                 78                5.50   
...                   ...                ...                 ...   
3301                156.8                103                7.06   
3304                206.5                 80                9.29   
3320                120.1                133                5.40   
3322                180.5                 72                8.12   
3323                227.0                 56               10.22   

      Total intl minutes  Total intl calls  Total intl charge  \
10                  12.7                 6               3.43   
15                   5.4                 9               1.46   
21                   5.7                 6               1.54   
33                  11.8                 3               3.19   
41                  14.6                15               3.94   
...                  ...               ...                ...   
3301                10.4                 4            

Вопрос: **Каковы средние значения числовых признаков среди нелояльных пользователей?**

👇

In [91]:
df[df['Churn'] == 1].mean()

Account length            102.664596
Area code                 437.817805
Number vmail messages       5.115942
Total day minutes         206.914079
Total day calls           101.335404
Total day charge           35.175921
Total eve minutes         212.410145
Total eve calls           100.561077
Total eve charge           18.054969
Total night minutes       205.231677
Total night calls         100.399586
Total night charge          9.235528
Total intl minutes         10.700000
Total intl calls            4.163561
Total intl charge           2.889545
Customer service calls      2.229814
Churn                       1.000000
dtype: float64

Скомбинировав предыдущие два вида индексации, ответим на вопрос: **Сколько в среднем в течение дня разговаривают по телефону нелояльные пользователи?**

👇

In [92]:
df[df['Churn'] == 1]['Total day minutes'].mean()

206.91407867494814

Вопрос: **Какова максимальная длина международных звонков среди лояльных пользователей, не пользующихся услугой международного роуминга?**

👇

In [93]:
df[(df['Churn'] == 0) & (df['International plan'] == 'No')]['Total intl minutes'].max()

18.9

DataFrame можно индексировать как по названию столбца или строки, так и по порядковому номеру. 

Для индексации **по названию** используется метод `loc`. В этом случае мы говорим "*передай нам значения для строк с id от 5 до 15 и для столбцов от State до Voice mail plan*".

In [94]:
df.loc[5:15, 'State':'Voice mail plan']

State  Account length  Area code International plan Voice mail plan
5     AL             118        510                Yes              No
6     MA             121        510                 No             Yes
7     MO             147        415                Yes              No
8     LA             117        408                 No              No
9     WV             141        415                Yes             Yes
10    IN              65        415                 No              No
11    RI              74        415                 No              No
12    IA             168        408                 No              No
13    MT              95        510                 No              No
14    IA              62        415                 No              No
15    NY             161        415                 No              No

Для индексации **по номеру** используется метод `iloc`. В этом случае мы говорим "*передай нам значения строк с 5 до 15 первых пяти столбцов*".

In [95]:
df.iloc[5:15, 0:5]

State  Account length  Area code International plan Voice mail plan
5     AL             118        510                Yes              No
6     MA             121        510                 No             Yes
7     MO             147        415                Yes              No
8     LA             117        408                 No              No
9     WV             141        415                Yes             Yes
10    IN              65        415                 No              No
11    RI              74        415                 No              No
12    IA             168        408                 No              No
13    MT              95        510                 No              No
14    IA              62        415                 No              No

Если нам нужна первая строчка DataFrame, используем конструкцию `df[:1]` (аналогично срезам для списков и строк).

In [96]:
df[:1]

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3                2.7   

   Customer service calls  Churn  
0                       1      0

Если нам нужна последняя строчка DataFrame, используем конструкцию `df[-1:]` (также аналогично срезам для списков и строк).

In [97]:
df[-1:]

State  Account length  Area code International plan Voice mail plan  \
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3332             39.85              265.9               82              22.6   

      Total night minutes  Total night calls  Total night charge  \
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3332                13.7                 4                3.7   

      Customer service calls  Churn  
3332                       0      0

### Применение функций к ячейкам, столбцам и строкам

**Применение функции к каждому столбцу: `apply()`**

К каждому столбцу применяется функция `max` из библиотеки `NumPy` для определения максимального элемента в столбце. Метод `apply()` можно использовать и для того, чтобы применить функцию к каждой строке. Для этого нужно указать параметр `axis=1` – `df.apply(np.max, axis=1)`.

In [98]:
df.apply(np.max)

State                        WY
Account length              243
Area code                   510
International plan          Yes
Voice mail plan             Yes
Number vmail messages        51
Total day minutes         350.8
Total day calls             165
Total day charge          59.64
Total eve minutes         363.7
Total eve calls             170
Total eve charge          30.91
Total night minutes         395
Total night calls           175
Total night charge        17.77
Total intl minutes           20
Total intl calls             20
Total intl charge           5.4
Customer service calls        9
Churn                         1
dtype: object

**Применение функции к каждой ячейке столбца: `map()`**

Метод `map()` можно использовать для **замены значений в колонке**, передав ему в качестве аргумента словарь вида `{old_value: new_value}`.

In [99]:
d = {'No': False, 'Yes': True}
df['International plan'] = df['International plan'].map(d)
df.head()

State  Account length  Area code  International plan Voice mail plan  \
0    KS             128        415               False             Yes   
1    OH             107        415               False             Yes   
2    NJ             137        415               False              No   
3    OH              84        408                True              No   
4    OK              75        415                True              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

Аналогичную операцию можно провернуть с помощью метода `replace`.

In [100]:
df = df.replace({'Voice mail plan': d})
df.head()

State  Account length  Area code  International plan  Voice mail plan  \
0    KS             128        415               False             True   
1    OH             107        415               False             True   
2    NJ             137        415               False            False   
3    OH              84        408                True            False   
4    OK              75        415                True            False   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

### Группировка данных

В общем случае группировка данных в Pandas выглядит следующим образом:

`df.groupby(by=grouping_columns)[columns_to_show].function()`

1. К датафрейму применяется метод `groupby`, который разделяет данные по `grouping_columns` – признаку или набору признаков.

2. Выбираются нужные столбцы (`columns_to_show`).

3. К полученным группам применяется функция или несколько функций.

**Группирование данных в зависимости от значения признака `Churn` и вывод статистик по трём столбцам в каждой группе.**

In [101]:
columns_to_show = ['Total day minutes', 
                   'Total eve minutes', 
                   'Total night minutes']

df.groupby(['Churn'])[columns_to_show].describe(percentiles=[])

Total day minutes                                            \
                  count        mean        std  min    50%    max   
Churn                                                               
0                2850.0  175.175754  50.181655  0.0  177.2  315.6   
1                 483.0  206.914079  68.997792  0.0  217.6  350.8   

      Total eve minutes                                             \
                  count        mean        std   min    50%    max   
Churn                                                                
0                2850.0  199.043298  50.292175   0.0  199.6  361.8   
1                 483.0  212.410145  51.728910  70.9  211.3  363.7   

      Total night minutes                                              
                    count        mean        std   min     50%    max  
Churn                                                                  
0                  2850.0  200.133193  51.105032  23.2  200.25  395.0  
1                   483.0  205.231677  47.132825  47.4  204.80  354.9

То же самое можно сделать, передав в функцию `agg()` (одним из аргументов принимает одну или несколько функций, которые нужно применить к заданной абсциссе) список функций.

In [102]:
df.groupby(['Churn'])[columns_to_show].agg([np.mean, np.std, np.min, np.max])

Total day minutes                        Total eve minutes             \
                   mean        std amin   amax              mean        std   
Churn                                                                         
0            175.175754  50.181655  0.0  315.6        199.043298  50.292175   
1            206.914079  68.997792  0.0  350.8        212.410145  51.728910   

                   Total night minutes                          
       amin   amax                mean        std  amin   amax  
Churn                                                           
0       0.0  361.8          200.133193  51.105032  23.2  395.0  
1      70.9  363.7          205.231677  47.132825  47.4  354.9

### Сводные таблицы

Допустим, необходимо посмотреть, как наблюдения в выборке распределены в контексте двух признаков – `Churn` и `International plan`. Для этого можно построить **таблицу сопряженности**, воспользовавшись методом `crosstab()`.

In [103]:
pd.crosstab(df['Churn'], df['International plan'])

International plan  False  True 
Churn                           
0                    2664    186
1                     346    137

In [104]:
pd.crosstab(df['Churn'], df['International plan'], normalize=True)

International plan    False     True 
Churn                                
0                   0.79928  0.055806
1                   0.10381  0.041104

В Pandas, так же, как и в Exel, есть такая опция, как **сводные таблицы** (*pivot tables*). В Pandas за сводные таблицы отвечает метод `pivot_table`, который принимает в качестве параметров:

+ `values` – список переменных, по которым требуется рассчитать нужную статистику,

+ `index` – список переменных, по которым нужно сгрупировать данные,

+ `aggfunc` – то, что необходимо посчитать по группам (сумму, среднее, максимум, минимум и др.)

Выведем среднее число дневных, вечерних и ночных звонков для разных Area code:

In [105]:
df.pivot_table(['Total day calls', 'Total eve calls', 'Total night calls'], ['Area code'], aggfunc='mean')

Total day calls  Total eve calls  Total night calls
Area code                                                     
408             100.496420        99.788783          99.039379
415             100.576435       100.503927         100.398187
510             100.097619        99.671429         100.601190

### Преобразование датафреймов

Добавить столбцы в DataFrame можно несколькими способами.

Например, необходимо посчитать общее количество звонков для всех пользователей. Для этого создадим объект `total_calls` типа Series и вставим его в датафрейм.

Метод `insert()` вставляет столбец в DataFrame в указанном месте. В данном примере, вставка осуществляется в конец таблицы (`loc=len(df.columns)`).

In [106]:
total_calls = df['Total day calls'] + df['Total eve calls'] + \
              df['Total night calls'] + df['Total intl calls']

df.insert(loc=len(df.columns), column='Total calls', value=total_calls)
df

State  Account length  Area code  International plan  Voice mail plan  \
0       KS             128        415               False             True   
1       OH             107        415               False             True   
2       NJ             137        415               False            False   
3       OH              84        408                True            False   
4       OK              75        415                True            False   
...    ...             ...        ...                 ...              ...   
3328    AZ             192        415               False             True   
3329    WV              68        415               False            False   
3330    RI              28        510               False            False   
3331    CT             184        510                True            False   
3332    TN              74        415               False             True   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  ...  Total eve charge  \
0                45.07              197.4  ...             16.78   
1                27.47              195.5  ...             16.62   
2                41.38              121.2  ...             10.30   
3                50.90               61.9  ...              5.26   
4                28.34              148.3  ...             12.61   
...                ...                ...  ...               ...   
3328             26.55              215.5  ...             18.32   
3329             39.29              153.4  ...             13.04   
3330             30.74              288.8  ...             24.55   
3331             36.35              159.6  ...             13.57   
3332             39.85              265.9  ...             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6  

Добавить столбец можно более простым способом, без создания промежуточных Series – с помощью присвоения:

In [107]:
df['Total charge'] = df['Total day charge'] + df['Total eve charge'] + \
                     df['Total night charge'] + df['Total intl charge']
df

State  Account length  Area code  International plan  Voice mail plan  \
0       KS             128        415               False             True   
1       OH             107        415               False             True   
2       NJ             137        415               False            False   
3       OH              84        408                True            False   
4       OK              75        415                True            False   
...    ...             ...        ...                 ...              ...   
3328    AZ             192        415               False             True   
3329    WV              68        415               False            False   
3330    RI              28        510               False            False   
3331    CT             184        510                True            False   
3332    TN              74        415               False             True   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  ...  Total night minutes  \
0                45.07              197.4  ...                244.7   
1                27.47              195.5  ...                254.4   
2                41.38              121.2  ...                162.6   
3                50.90               61.9  ...                196.9   
4                28.34              148.3  ...                186.9   
...                ...                ...  ...                  ...   
3328             26.55              215.5  ...                279.1   
3329             39.29              153.4  ...                191.3   
3330             30.74              288.8  ...                191.9   
3331             36.35              159.6  ...                139.2   
3332             39.85              265.9  ...                241.4   

      Total night calls  Total night charge  Total intl minutes  \
0                    91               11.01                10.0   
1                   103               11.45                13.7   
2                   104                7.32                12.2   
3                    89                8.86                 6.6   
4                   121                8.41                10.1   
...                 ...                 ...                 ...   
3328                 83               12.56                 9.9   
3329                123                8.61                 9.6   
3330                 91                8.64                14.1   
3331                137                6.26                 5.0   
3332                 77               10.86                13.7   

      Total intl calls  Total intl charge  Customer service calls  Churn  \
0                    3               2.70                       1      0   
1                    3               3.70                       1      0   
2                    5               3.29                       0      0   
3                    7               1.78                       2      0   
4                    3               2.73                       3      0   
...                ...                ...                     ...    ...   
3328                 6               2.67                       2      0 

Для удаления столбцов или строк используется метод `drop()`. В качестве аргументов передаются нужные индексы (списком) и требуемое значение параметра `axis` (`1` – для удаления столбца (признака), `0` – для удаления строки (значение по умолчанию)).

Удалим созданные только что столбцы:

In [108]:
df = df.drop(['Total charge', 'Total calls'], axis=1)
df

State  Account length  Area code  International plan  Voice mail plan  \
0       KS             128        415               False             True   
1       OH             107        415               False             True   
2       NJ             137        415               False            False   
3       OH              84        408                True            False   
4       OK              75        415                True            False   
...    ...             ...        ...                 ...              ...   
3328    AZ             192        415               False             True   
3329    WV              68        415               False            False   
3330    RI              28        510               False            False   
3331    CT             184        510                True            False   
3332    TN              74        415               False             True   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9      

Строки удаляются следующим образом:

In [109]:
df.drop([1, 2]).head()

State  Account length  Area code  International plan  Voice mail plan  \
0    KS             128        415               False             True   
3    OH              84        408                True            False   
4    OK              75        415                True            False   
5    AL             118        510                True            False   
6    MA             121        510               False             True   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
3                      0              299.4               71   
4                      0              166.7              113   
5                      0              223.4               98   
6                     24              218.2               88   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   
5             37.98              220.6              101             18.75   
6             37.09              348.5              108             29.62   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   

   Customer service calls  Churn  
0                       1      0  
3                       2      0  
4                       3      0  
5                       0      0  
6                       3      0

### Прогнозирование оттока

Посмотрим, как отток связан с признаком "*International plan*". Используем сводную таблицу `crosstab`. При значении аргумента `margins=True` в таблицу добавляются результаты суммирования элементов столбцов, строк и всей таблицы в целом.

In [113]:
pd.crosstab(df['Churn'], df['International plan'], margins=True)

International plan  False  True   All
Churn                                
0                    2664   186  2850
1                     346   137   483
All                  3010   323  3333

Видно, что когда роуминг подключен, доля оттока намного выше. Возможно, большие и плохо контролируемые траты в роуминге очень не нравятся клиентам и приводят к их недовольству телеком-оператором и, соответственно, к их оттоку. 

Посмотрим на ещё один важный признак – "*Customer service calls*". Построим сводную таблицу.

In [116]:
pd.crosstab(df['Churn'], df['Customer service calls'], margins=True)

Customer service calls    0     1    2    3    4   5   6  7  8  9   All
Churn                                                                  
0                       605  1059  672  385   90  26   8  4  1  0  2850
1                        92   122   87   44   76  40  14  5  1  2   483
All                     697  1181  759  429  166  66  22  9  2  2  3333

Видно, что доля оттока сильно возрастает начиная с 4х звонков в сервисный центр (смотреть на соотношение значений в столбцах).

Теперь добавим в DataFrame бинарный признак – результат сравнения `Customer service calls > 3`:

In [117]:
df['Many service calls'] = (df['Customer service calls'] > 3).astype('int')

pd.crosstab(df['Many service calls'], df['Churn'], margins=True)

Churn                  0    1   All
Many service calls                 
0                   2721  345  3066
1                    129  138   267
All                 2850  483  3333

Объединим рассмотренные выше условия и построим сводную таблицу для этого объединения и оттока.

In [118]:
pd.crosstab(df['Many service calls'] & df['International plan'], df['Churn'])

Churn     0    1
row_0           
False  2841  464
True      9   19

Таким образом, прогнозируя отток клиентов в случае, когда число звонков в сервисный центр превышало 3 и подключен роуминг (и прогнозируя лояльность – в противном случае), можно ожидать около 85.8% правильных попаданий (ошибка всего в 464 + 9 случаях). Полученные 85.8% неплохая отправная точка (*baseline*) для дальнейшего построения моделей машинного обучения. 

### Резюме:

+ Доля лояльных клиентов в выборке – 85.5%. Самая наивная модель, ответ которой "клиент всегда лоялен" на подобных данных будет угадывать примерно в 85.5% случаев. То есть доли правильных ответов (*accuracy*) последующих моделей должны быть как минимум не меньше, а лучше – значительно выше этого значения;

+ С помощью простого прогноза, который условно можно выразить следующей формулой: `(International plan = True) & (Customer service calls > 3) => Churn = 1, else Churn = 0`, можно ожидать долю угадываний 85.8%, что ещё чуть выше 85.5%.

+ Эти две отправные точки (*baselines*) были получены без всякого машинного обучения. Если окажется, что прикладываемыми усилиями мы увеличиваем долю правильных ответов всего, скажем, на 0.5%, то возможно, мы что-то делаем не так, и достаточно ограничиться простой моделью из двух условий.

+ Перед обучением сложных моделей рекомендуется немного покрутить данные и проверить простые предположения. Более того, в бизнес-приложениях машинного обучения чаще всего начинают именно с простых решений, а потом экспериментируют с их усложнениями. 
